WalletWasabi/WabiSabi/Client/CredentialDependencies/DependencyGraph.cs

```diff
@@ -0,0 +1,443 @@
+				{
+					// TODO shuffle? not clear if it will benefit or reduce
+					// privacy, need to study it in more detail. The adverserial
+					// setting that should be considered is one where the
+					// coordinator can delay information and request processing
+					// and perform an arbitrary number of blameless round
+					// failures while attempting to detect correlations between
+					// the timings of the registrations of a single user. Random
+					// delays should add noise to this and ideally all clients
+					// should aim to have the same fixed baseline delay for
+					// their output registrations (something that should almost
+					// always be a few times longer than the expected RTT), then
+					// output registrations should mostly follow reissuances and
+					// be difficult to tell apart based on timing, which should
+					// render the ordering irrelevant.
+
+					if (negative.Count() > unconstrainedPositive.Count())
+					{
+						// TODO consolidate only lowest depth at each iteration
+						// (negative height, i.e. max of distance to terminal output
+						// nodes derived from the vertex, 0 for such nodes) nodes
+						// first. By only consolidating nodes with the lowest
+						// depth a balanced structure will be ensured regardless
+						// of prior reissuances potentially creating deeper
+						// nodes.
+						(g, negative) = g.ReduceNodes(negative, unconstrainedPositive.Count(), credentialType);
+					}
+				}
+			}
+
+			edgeSet = g.edgeSets[credentialType];
+
+			// Second special case, more general to the previous one, when for each
+			// negative node there is a satisfactory unconstrained positive node
+			// (not necessarily all of equal value), discharge via a 1:1 correspondence
+			if (negative.Count() <= unconstrainedPositive.Count() && Enumerable.Zip(unconstrainedPositive, negative).All(p => edgeSet.Balance(p.First) + edgeSet.Balance(p.Second) >= 0))
+			{
+				g = Enumerable.Zip(unconstrainedPositive, negative).Aggregate(g, (g, p) => g.DrainTerminal(p.First, new RequestNode[] { p.Second }, credentialType));
+			}
+
+			return g;
+		}
+
+		// Build a k-ary tree bottom up to reduce a list of nodes to discharge
+		// to at most maxCount elements.
+		private (DependencyGraph, IEnumerable<RequestNode>) ReduceNodes(IEnumerable<RequestNode> nodes, int maxCount, CredentialType credentialType)
+		{
+			if (nodes.Count() <= maxCount)
+			{
+				return (this, nodes);
+			}
+
+			// Replace up to k nodes, possibly the entire queue, with a
+			// single reissuance node which combines their values. The total
+			// number of items might be less than K but still larger than
+			// maxCount.
+			var take = Math.Min(K, nodes.Count());
+			(var g, var reissuance) = AggregateIntoReissuanceNode(nodes.Take(take), credentialType);
+			var reduced = nodes.Skip(take).Append(reissuance).ToImmutableArray(); // keep enumerable expr size bounded by evaluating eagerly
```

Comment:
`ToImmutableArray` seems to be unnecessary.

Reply:
:+1:, i think there used to be a reordering or something but `Skip` should be O(1), and both `Skip` and `Append` are called log(n) times in total.

I wonder how clever `Skip` is at pruning the layers of `Append` that can't contribute elements, assuming `node` has a known count. I imagine none of this is covered by any spec, but are there resources for what sort of performance to expect out of linq in circumstances like this?

Reply:
ah, note that the first `nodes` in the recursion will (usually) be one of these. IIRC this `ToImmutableArray` and comment was added when that was performing `O(n log^2 n)` work sorting, when I reified it there I didn't remove it here. For what it's worth, it did seem to make a noticeable performance difference, but I didn't actually try to measure how much.
 https://github.com/zkSNACKs/WalletWasabi/blob/92d312e5a53f85a23e97b3373dc067db70f7545f/WalletWasabi/WabiSabi/Client/CredentialDependencies/CredentialEdgeSet.cs#L126-L135